feat: add "auto" variant for 2023 icons

arthur-fontaine opened the pull request:
Since VSCode 1.42 (January 2020), VSCode supports automatic theme switching based on the system color scheme (see [release notes](https://code.visualstudio.com/updates/v1_42#_auto-switch-theme-based-on-os-color-scheme)).

Since the light icons do not match the dark themes, and the dark icons do not match the light themes, it may be useful to provide a variant called "Auto" that automatically switches to light or dark icons depending on the system color scheme.

I created a script that is automatically run when the extension is built (NPM automatically run `prebuild` script when the `build` script is run).

https://github.com/arthur-fontaine/vscode-jetbrains-icon-theme/blob/f5895c956612fb3eff6a8cd9b009a3dbc6e191f2/package.json#L13

This script will go through all the 2023 icons, pick the light and dark (if exists) for each one, and create a new SVG file with a group for the light icon and a group for the dark icon. These groups are displayed or not thanks to a media query in CSS (`prefers-color-scheme`). The script will also generate a new `theme-auto.json` file, based on `theme-light.json`.

Demo (look at the icons, e.g. the JS and the folder icons change colors):


https://github.com/cadamsdev/vscode-jetbrains-icon-theme/assets/57626613/3d53d5d9-e171-4d7a-9b30-42de527a695b

cadamsdev commented:
Hey! Thanks for contributing. 🙂

A couple things I would like to look into before we merge this.
1. Can we look into an approach where we don't have to commit the auto svg files?
2. Instead of having 3 icon themes for "JetBrains Icon Theme 2023+" we could remove the [light](https://github.com/cadamsdev/vscode-jetbrains-icon-theme/blob/f5895c956612fb3eff6a8cd9b009a3dbc6e191f2/package.json#L53) and [dark](https://github.com/cadamsdev/vscode-jetbrains-icon-theme/blob/f5895c956612fb3eff6a8cd9b009a3dbc6e191f2/package.json#L57) then just name this new icon theme "JetBrains Icon Theme 2023+ UI"?

arthur-fontaine commented:
(1) Yes. I think that we could rename the current `themes` folder to an `assets` folder. Then we would have a `scripts/build.sh` script that packages the icons into a `dist` folder. This script is run by the `prebuild` NPM script.

(2) I think this is a bad idea. I think that a lot of people are still not using an adaptive theme (they still have a light theme but their computers are in dark mode, or vice versa). Since the icons are themed according to the OS color scheme, that would not fit for some people. So I think that we need to keep 3 icon themes.
However, we could make the auto one the default?

cadamsdev commented:
> (1) Yes. I think that we could rename the current `themes` folder to an `assets` folder. Then we would have a `scripts/build.mjs` script that packages the icons into a `dist` folder. This script is run by the `prebuild` NPM script.

Sounds good to me.

> (2) I think this is a bad idea. I think that a lot of people are still not using an adaptive theme (they still have a light theme but their computers are in dark mode, or vice versa). Since the icons are themed according to the OS color scheme, that would not fit for some people. So I think that we need to keep 3 icon themes. However, we could make the auto one the default?

That's fair. I'm ok with that. Feel free to make those changes.

arthur-fontaine commented:
Just added the new `build` script, renamed the `themes` folder to `assets` and changed the theme paths in the `package.json` to point to `dist...`. Also rebased.

I let you version the package.

cadamsdev commented on scripts/build/generator/generate-2023-auto-theme.mjs:
```diff
@@ -0,0 +1,166 @@
+        icons[iconDefinitionKey][iconDefinition.theme] = iconAst
+    }
+
+    return /** @type {any} */ (icons)
```
I would recommend using TypeScript instead of typing with JSDoc. JSDoc is better suited for when you're building a library.

Also this may be subjective but personally not a big fan of the JSDoc syntax. TypeScript seems much cleaner and easier to  read.

If you convert this to TypeScript we can use the [ts-node](https://www.npmjs.com/package/ts-node) package to run the typescript file on `prebuild`

arthur-fontaine replied:
I used `/** @type {any} */` to prevent VSCode from marking an error and because I am sure that it matches the required return type. It can be safely removed.

I did not want to not add new dependencies to run Typescript (it is not really necessary), but I wanted some type safety because the functions are a bit complex.

I don't think this is a problem, but if you really think it is, we can also remove the JSDoc comments. If you want to use Typescript, I would recommend using [tsx](https://github.com/esbuild-kit/tsx) (see the [comparisons](https://github.com/privatenumber/ts-runtime-comparison) with ts-node) to run the build script.

cadamsdev replied:
If you don't mind using TypeScript then let's use that. Feel free to use tsx though.

cadamsdev commented on scripts/build/types/Theme.d.ts:
```diff
@@ -0,0 +1,8 @@
+export interface Theme {
```
If we use TypeScript, TypeScript will automatically generate the type declaration files for us. This seems tedious having to manually create these.

cadamsdev commented on scripts/build/utils/build-theme.ts:
```diff
@@ -0,0 +1,31 @@
+import * as fs from "fs";
```
To be consistent with
https://github.com/cadamsdev/vscode-jetbrains-icon-theme/pull/103/files#diff-7d51ff9d2b2fe28ba3bad3e3a41dddf848d80a6a7e5a184d16317a00b980733dR3

```suggestion
import * as fs from "node:fs";
```

cadamsdev commented on scripts/build/utils/build-theme.ts:
```diff
@@ -0,0 +1,31 @@
+import * as fs from "fs";
+import * as path from "path";
```
To be consistent with
https://github.com/cadamsdev/vscode-jetbrains-icon-theme/pull/103/files#diff-9f1ca9806327e5685f7a6e6528cfa7ccea6b043cc3bde415cdffec38ccfbbec3R1

```suggestion
import * as path from "node:path";
```

cadamsdev commented on scripts/build/build.ts:
```diff
@@ -0,0 +1,44 @@
+import * as fs from "node:fs";
+import * as path from "node:path";
+
+import { SRC_DIR_PATH, BUILD_DIR_PATH } from "./constants.js";
```
```suggestion
import { SRC_DIR_PATH, BUILD_DIR_PATH } from "./constants";
```

cadamsdev commented on scripts/build/build.ts:
```diff
@@ -0,0 +1,44 @@
+
+import { SRC_DIR_PATH, BUILD_DIR_PATH } from "./constants.js";
+import { generate2023AutoTheme } from "./generators/generate-2023-auto-theme";
+import { buildTheme } from "./utils/build-theme.js";
```
```suggestion
import { buildTheme } from "./utils/build-theme";
```

arthur-fontaine commented:
Everything fixed + fix conflicts ✅

cadamsdev commented on scripts/build/utils/get-icon-paths.ts:
```diff
@@ -0,0 +1,9 @@
+
+import { Theme } from "../types/Theme";
+
+export function getIconPaths(theme: Theme, themePath = "") {
```
Add explicit type 
```suggestion
export function getIconPaths(theme: Theme, themePath = ""): string[] {
```

cadamsdev commented:
Looks good to me. Thanks for contributing. 🙂

arthur-fontaine commented:
Thank you for the collaboration 😉

arthur-fontaine commented:
<img width="451" alt="Capture d’écran 2023-11-02 à 01 09 33" src="https://github.com/cadamsdev/vscode-jetbrains-icon-theme/assets/57626613/b4150504-e47a-4570-9d4b-4af5c6b2780c">
<img width="871" alt="Capture d’écran 2023-11-02 à 01 09 23" src="https://github.com/cadamsdev/vscode-jetbrains-icon-theme/assets/57626613/3ebfe1c1-e4d7-44e7-926f-2d859c14c1f3">

Seems like the extension is not built correctly :/.

cadamsdev commented:
> <img alt="Capture d’écran 2023-11-02 à 01 09 33" width="451" src="https://user-images.githubusercontent.com/57626613/279829522-b4150504-e47a-4570-9d4b-4af5c6b2780c.png"> <img alt="Capture d’écran 2023-11-02 à 01 09 23" width="871" src="https://user-images.githubusercontent.com/57626613/279829527-3ebfe1c1-e4d7-44e7-926f-2d859c14c1f3.png">
> Seems like the extension is not built correctly 🫤.

Looks like the [prebuild](https://github.com/cadamsdev/vscode-jetbrains-icon-theme/blob/main/package.json#L13) script is not being called.

I think we may need to call it before [here](https://github.com/cadamsdev/vscode-jetbrains-icon-theme/blob/main/.github/workflows/CI.yml#L17)
